### DISCUSSION TITLE
What are the closest preprocessors to DASM's DS and SEG preprocessors

### DISCUSSION THREAD

**Discussion by Pokeman2003** [vhelin/wla-dx#545] [Q&A] [answered]
Hi! I'm currently trying to program for the Atari 2600, and I'm struggling pretty hard. All of the documentation seems to exist for DASM and not WLA-DX. There's one file in particular that seems very important to port, which is [vcs.h](https://github.com/munsie/dasm/blob/master/machines/atari2600/vcs.h). 

The problem that I'm running into is that I can't seem to get a grip on both DASM and WLA-DX's preprocessors. Furthermore, I can't even quite figure out what a segment is supposed to be according to DASM, and its associated behavior. This is particularly bad because this file makes heavy use of DS and SEG. For a quick overview:
- DS:Declare space and fill with a fillvalue (if specified, otherwise default is 0).
- SEG:This switches to a new segment, creating it if necessary.
(Source:[Pages 46 and 54](https://raw.githubusercontent.com/dasm-assembler/dasm/master/docs/dasm.pdf))

Any help is greatly appreciated! TIA!

**Comment by vhelin** [comment]
Looking at vcs.h I think you'll need to completely rewrite it for WLA DX. That SEG(ment) looks like WLA's .SECTION

```
                        SEG.U TIA_REGISTERS_WRITE
			ORG TIA_BASE_WRITE_ADDRESS
	; DO NOT CHANGE THE RELATIVE ORDERING OF REGISTERS!
    
VSYNC       ds 1    ; $00   0000 00x0   Vertical Sync Set-Clear
VBLANK		ds 1	; $01   xx00 00x0   Vertical Blank Set-Clear
WSYNC		ds 1	; $02   ---- ----   Wait for Horizontal Blank
RSYNC		ds 1	; $03   ---- ----   Reset Horizontal Sync Counter
```

I think in WLA parlance this would be something like

```
.ORG TIA_BASE_WRITE_ADDRESS
.SECTION "TIA_REGISTERS_WRITE" FORCE
VSYNC:       db 0  ; allocates a byte and sets it to 0
VBLANK:     db 0
WSYNC:      db 0
RSYNC:       db 0
...
.ENDS
```

Then again I'm not totally sure if you want to allocate space there or would an .ENUM be enough?

```
.ENUM TIA_BASE_WRITE_ADDRESS
VSYNC db
VBLANK db
WSYNC db
RSYNC db
...
.ENDE
```

**Reply by Pokeman2003** [reply]
Hmm, I'll have to do some experimentation tomorrow when I can get my hands on a working computer. As a replacement, I just kind of made a file that had nothing BUT .DEFINEs and addresses.

If it helps, they're all defining memory addresses right before RAM(the registers to manipulate the machine, effectively.) That's why they're defining space there. Enum might work better though.

**Comment by vhelin** [comment, marked as answer]
Yeah, well, if VSYNC, VBLANK etc. are hardware registers that have addresses, but you don't want to put there data by default, only using your code, then .ENUM or .DEFINEs are what you need.

Gotta admit that I don't know anything about DASM or Atari 2600. :) But Amiga has hardware registers like Gameboy, somewhere in the address space, and those are just addresses/locations that can be defined using .DEFINEs or .ENUM and then written into in your code...

**Reply by Pokeman2003** [reply]
Both are valid ways to handle this. I personally prefer the .enum method, but at this point I don't want to rewrite the definitions file I made. Thanks!
